Mage/src/main/java/mage/abilities/keyword/KickerAbility.java

```diff
@@ -119,32 +114,30 @@ private void resetKicker() {
         for (OptionalAdditionalCost cost : kickerCosts) {
             cost.reset();
         }
-        activations.clear();
     }
 
-    private int getKickedCounterStrict(Game game, Ability source, String needKickerCost) {
-        String key;
-        if (needKickerCost.isEmpty()) {
-            // need all kickers
-            key = getActivationKey(source, "", game);
-        } else {
-            // need only cost related kickers
-            key = getActivationKey(source, needKickerCost, game);
-        }
+    private static String getActivationKey(String needKickerCost){
+        return "kickerActivation"+needKickerCost;
+    }
 
-        int totalActivations = 0;
-        if (kickerCosts.size() > 1) {
-            for (String activationKey : activations.keySet()) {
-                if (activationKey.startsWith(key) && activations.get(activationKey) > 0) {
-                    totalActivations += activations.get(activationKey);
-                }
-            }
-        } else {
-            if (activations.containsKey(key) && activations.get(key) > 0) {
-                totalActivations += activations.get(key);
-            }
+    /**
+     * Return total kicker activations with the specified Cost (blank for all kickers/multikickers)
+     * Checks the start of the tags, to work for that blank method, which requires direct access
+     *
+     * @param game
+     * @param source
+     * @param needKickerCost use cost.getText(true)
+     * @return
+     */
+    public static int getKickedCounterStrict(Game game, Ability source, String needKickerCost) {
+        Map<String, Object> costsTags = CardUtil.getSourceCostsTagsMap(game, source);
+        if (costsTags == null) {
+            return 0;
         }
-        return totalActivations;
+        String finalActivationKey = getActivationKey(needKickerCost);
+        Stream<Map.Entry<String, Object>> tagStream = costsTags.entrySet().stream()
+                .filter(x -> x.getKey().startsWith(finalActivationKey));
+        return tagStream.mapToInt(x -> (int)x.getValue()).sum();
```

Comment:
Miss null protection for getValue, aslo must use (Integer) instead (int). Must use same code style for all usages. As example: null values in tags for another ability:

![shot_231117_115259](https://github.com/magefree/mage/assets/8344157/b80a2404-722a-4982-b61d-5c4d21feabaa)

Reply:
If an object of the wrong type gets into the map with a key starting with "kickerActivation" (or with an exact key of another ability/"X") then a lot of things can go very wrong. As such I don't really think it's worth trying to null check it, better to fail loud and fast than to have silent errors. If no keys are found then the stream will be empty and `sum()` will return zero.

As I mentioned in the main text, Kicker is unique because a lot of existing cards check for the existence of any kicker activation while others check for kicker activations with a specific cost.

Also, mapToInt converts to an `int` internally anyway.

Reply:
1. `(int) null` will raise error before go to mapToInt
2. (Integer) null will silent and goes fine to mapToInt
3. (Integer) bad_class will raise error all time

Reply:
Potential bug example (there were many bugs with same style with watchers/conditionals):
- add card hint to show ability status;
- game will check it before any cast;
- it catch empty value and broke the game;

Reply:
The conversion is only after the key is already checked in the filter, so I don't see how null checking here (after the key's existence is confirmed) is useful. There must never be a situation where a key exists but the value stored with that key is of the wrong type. If that happens then it means either:
1. code is storing a different type from what it's trying to read with its key
2. there's multiple areas trying to store different values in the same key

In either situation we would want an immediate crash rather than a silent mishandling of data, correct?

Reply:
> There must never be a situation where a key exists but the value stored with that key is of the wrong type

See above -- you adds `null` value in it:
![1](https://user-images.githubusercontent.com/8344157/283720953-b80a2404-722a-4982-b61d-5c4d21feabaa.png)


So follow simple rules:
* if tag system allows null values then process null in all code (recommended);
* if tag system restrict null values then process it as string or raise exception on modifying;

Reply:
> Potential bug example (there were many bugs with same style with watchers/conditionals):
> 
>     * add card hint to show ability status;
> 
>     * game will check it before any cast;
> 
>     * it catch empty value and broke the game;

This is handled fine. Either the key does not exist and the default value is returned (or for Kicker the stream is empty), or the key does exist and the value must be of the correct type.

Reply:
> See above -- you adds `null` value in it:

Yes, with a key of `"entwineActivation"`. The String keys must be kept separate between different abilities. Values stored will be of type `Integer`, `null`, `HashSet<MageObjectReference>` for Convoke, and potentially others in the future. If the Kicker code tries to read the value stored by Entwine, something very major has gone wrong.

All values with a key of `entwineActivation` are null (since the key's existence is what is being checked). All `kickerActivation<COST>` values are Integer, and never null. Each key must have one type of object used inside of it.

If you really want me to add null checking everywhere I guess I can, but the design intent of the costs tags system very strongly does not need it and the only thing it could do would be hide an incorrect usage of the system.

Reply:
Yes, please, use `(Integer)` and null protection code all around (here and in goblin PR too). If you want to catch a wrong usage of kicker or other ability then add runtime check with exception error like "Wrong code usage: kicker miss activate value".

Reply:
> if tag system restrict null values then process it as string or raise exception on modifying;

To be clear, it's not just that it restricts *null* values: it wants *all* values for a given key to be of the same type. There is not currently enforcement of this restriction beyond the potential failed cast errors. If you have a suggestion on how to reasonably enforce this, I'm interested.

The only way I can currently think of would be to have a separate type-checking map that stores all used keys and the Class type of the values used by them. I could add that if you think it's worthwhile, but I think that the complexity of that system would be worse than just leaving it as-is.

Reply:
> Yes, please, use `(Integer)` and null protection code all around (here and in goblin PR too). If you want to catch a wrong usage of kicker or other ability then add runtime check with exception error like "Wrong code usage: kicker miss activate value".

Ok, thinking about it a bit more I can use the required default object and generics to handle the checking in there. It's an exception either way, but it's a more detailed IllegalStateException instead of just a standard type cast failure. For Kicker I can check `instanceof Integer` directly on all the values and throw an exception there. Since null is not an instance of integer, that does a null check there as well.

I hope this makes the intent more clear. If you for some reason still want me to allow incorrect nulls I will, but again the **only** thing that does would be making mistakes in code usage harder to notice.